pkg/backporter/httpcache.go

```diff
@@ -0,0 +1,146 @@
+package backporter
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"net/http"
+	"net/http/httputil"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	xFromCache       = "X-From-Cache"
+	cacheRefreshFreq = 10 * time.Minute
+)
+
+type bugzillaCache struct {
+	lock  sync.Mutex
+	cache map[string][]byte
+}
+
+func (bc *bugzillaCache) get(key string) ([]byte, bool) {
+	bc.lock.Lock()
+	defer bc.lock.Unlock()
+	cachedVal, ok := bc.cache[key]
+	return cachedVal, ok
+}
+
+func (bc *bugzillaCache) set(key string, respBytes []byte) {
+	bc.lock.Lock()
+	defer bc.lock.Unlock()
+	bc.cache[key] = respBytes
+}
+
+func (bc *bugzillaCache) delete(key string) {
+	bc.lock.Lock()
+	defer bc.lock.Unlock()
+	delete(bc.cache, key)
+}
+func newBugzillaCache() *bugzillaCache {
+	return &bugzillaCache{cache: map[string][]byte{}}
+}
+
+// cachingTransport is an implementation http.RoundTripper
+// which first checks for cached values
+type cachingTransport struct {
+	cache     *bugzillaCache
+	transport http.RoundTripper
+}
+
+// RoundTrip will first check if there are any cached responses and return that
+// if not it will make an HTTP call using the DefaultTransport
+func (t *cachingTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	if req.Method != "GET" {
+		return t.transport.RoundTrip(req)
+	}
+	var err error
+	var resp *http.Response
+	g := new(errgroup.Group)
+	g.Go(func() error {
+		resp, err = t.transport.RoundTrip(req)
+		if err != nil {
+			t.cache.delete(req.URL.String())
+			return err
+		}
+		body, err := httputil.DumpResponse(resp, true)
+		if err != nil {
+			t.cache.delete(req.URL.String())
+			return fmt.Errorf("err while prepping response to cache: %w", err)
+		}
+		if resp.StatusCode == http.StatusOK || resp.StatusCode == http.StatusNotFound {
+			t.cache.set(req.URL.String(), body)
+		} else {
+			t.cache.delete(req.URL.String())
```

Comment:
Deleting this here is wrong. Just keep whats in the cache, requests can fail for all kind of random reasons, there is no reason to delete a cache entry for every unanticipated error we get. If the entry is actually gone, we cache that

Reply:
OK. At what point would I delete something from the cache if its no longer relevant?

Reply:
Never